### PR TITLE
new source data format, sources, new charts

### LINKED ISSUE #31: Sources and Methodology functionality
There are two parts to this task. Both are described in the "Sources and Methodology" section of the PRD ([link](https://docs.google.com/document/d/1eOiMzk8_eDjC4W0xn89qlkmj_6SOqi1dBTyEusAGGaI/edit?ts=5fb45edc#bookmark=id.4cqy16qlfvg9)). 

### Methodology 
Each “metric block” (in Key Insights and the Flow Diagram) should have a Methodology tooltip available that summarizes the fields used to calculate that particular metric. This should be accessible from a blue tooltip in the corner of each box, and should generally read: "Includes data for the following categories from [source]’s public reports: [comma-delimited mapping_overrides]" 
Both [source] and [comma-delimited mapping_overrides] will be available in the fixture files. @colincadams to provide more details here. 

### Sources
This will be a simple section at the bottom of the page with a list of data sources for the state. The list of sources (URLs) will also be available in the fixture file. (@colincadams to provide details)
See PRD for mocks.

### LINKED ISSUE #32: New graphs for parole and probation revocations
We need to add two new graphs to show trends for parole and probation revocations when available. 

These will operate in exactly the same way as the existing graphs on the page. They should have the following data scope: 

**Parole Revocations**
- Total 
- New Crime
- Technical Violation

**Probation Revocations**
- Total 
- New Crime
- Technical Violation

This is how they should appear (data fields in fixture file are in the blue boxes): 

Screen Shot 2021-01-19 at 10 10 46 PM

### LINKED ISSUE #36: Remap metric names to match new scheme
**Is your feature request related to a problem? Please describe.**

I had mentioned this in #27 but wanted to file a separate issue to track the effort. In `src/constants/metrics.js` there are currently some mis-mapped metrics and some that need to be remapped based on the updated mocks: https://www.figma.com/file/66HyjH0efZMoAuTlhnojWh/Justice-Counts-Report?node-id=686%3A3462

**Describe the solution you'd like**

Update to match the new mocks. I think it should be straightforward to just copy them over to metrics.js, but if it is easier for me to enumerate the necessary changes let me know!

**Describe alternatives you've considered**

**Additional context**
Add any other context about the feature request here.

### PR DESCRIPTION
**Pull request by Wmaileq**
## Description of the change

- Replaced mock data with newer one
- Added 2 charts
- Added sources to Flow data and Key Insights cards
- Added Sources section

## Type of change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [X] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Configuration change (adjusts configuration to achieve some end related to functionality, development, performance, or security)

## Related issues

> Closes #31.
> Closes #32.
> Closes #36.

### Development

These boxes should be checked by the submitter prior to merging:

- [X] Manual testing against realistic data has been performed locally

### Code review

These boxes should be checked by reviewers prior to merging:

- [x] This pull request has a descriptive title and information useful to a reviewer
- [x] This pull request has been moved out of a Draft state, has no "Work In Progress" label, and has assigned reviewers
- [x] Potential security implications or infrastructural changes have been considered, if relevant

### CODE REVIEW

File: src/components/Sources/Sources.js [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+  }, []);
+
+  return (
+    <div className="Sources">
+      <h2 className="Sources__title">Sources</h2>
+      {!data.length ? (
+        <p className="Sources__no-data">No public sources available.</p>
+      ) : (
+        <p className="Sources__data">
+          All data for these visualizations was from public reports published by the {sourcesText}
```
Comment (jessex) [inline src/components/Sources/Sources.js:55]: We should add a period after `{sourcesText}`.

Comment (jessex) [reply src/components/Sources/Sources.js:55]: I think the text from the mocks reads a little strange. Let's change "was" to "comes".

File: src/utils/generateSourceData.js [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+// =============================================================================
+const generateSourceData = (flowData) => {
```
Comment (jessex) [inline src/utils/generateSourceData.js:17]: Let's make sure to add a docstring that explains the expected output format.

File: src/components/Sources/Sources.js [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+// =============================================================================
+import React from "react";
+import { sourcesPropTypes } from "./propTypes";
+
+import "./Sources.scss";
+
+const Sources = ({ data }) => {
```
Comment (jessex) [inline src/components/Sources/Sources.js:22]: Let's make sure to add a docstring that explains the expected output format.

File: src/utils/getNormalizedStateData.js [hunk 4]
```diff
@@ -30,6 +30,9 @@ import sortByYearAndMonth from "./sortByYearAndMonth";
  *   compared_to_month?: string
  *   value_change?: number
  *   percentage_change?: number
+ *   source_name: string
```
Comment (jessex) [inline src/utils/getNormalizedStateData.js:33]: We also want to update the `README` to thoroughly document the updated format of the data file.

File: src/components/Sources/Sources.js [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    acc.push(` ${name} (`);
+
+    links.forEach((link, linkIndex) => {
+      acc.push(
+        <a target="_blank" rel="noreferrer" className="Sources__link" href={link}>
+          link {linkIndex + 1}
+        </a>
+      );
+      if (linkIndex !== links.length - 1) {
+        acc.push(",");
```
Comment (jessex) [inline src/components/Sources/Sources.js:39]: This should be `", "`, I think, to ensure there's proper spacing after the comma. For example, we want a space after the comma here:

Screen Shot 2021-01-28 at 14 02 26

**Review by jessex — CHANGES_REQUESTED** [review]
Almost there, @Wmaileq! I identified a handful of minor issues in testing, though:

**Edge cases / error handling**

This tooltip reads "undefined null" instead of a more graceful handling:
Screen Shot 2021-01-28 at 14 01 57

This text reads "NaN people" instead of more graceful handling:
Screen Shot 2021-01-28 at 14 05 50

**Mobile styling**

In mobile, these new tooltips on the _Key Insights_ cards are cut off by the element just above:
Screen Shot 2021-01-28 at 14 04 03

In mobile, the "Not available" text in the cards where data is not available is not centered properly:
Screen Shot 2021-01-28 at 14 04 51

In mobile, the spacing for the _Sources_ section at the bottom is generally a bit off, it runs all the way up to the edges, etc.:
Screen Shot 2021-01-28 at 14 05 29

**Errors in the Console**
Screen Shot 2021-01-28 at 14 44 30

File: src/components/Sources/Sources.js (hunk 3 above)
Comment (Wmaileq) [reply src/components/Sources/Sources.js:22]: @jessex separators (`,` and `and`) replaced from js realization to css `before` and `after`. I think it became clearer now.

File: README.md [hunk 6]
```diff
@@ -87,3 +87,6 @@ An _example_ of a valid `data.json` file can be found in [`data.json`](src/data.
 - `compared_to_month` - The month of the comparison data point, which is the most recent data point to compare against for trend analysis, no less than one year before this current data point, e.g. `04`
 - `value_change` - Metric value change between `compared_to_year/month` and `year/month`, e.g. `-20`
 - `percentage_change` - Metric value change between `compared_to_year/month` and `year/month` in percents, e.g. `-0.12`
+- `source_name` - Source provider name
```
Comment (jessex) [inline README.md:90]: Let's change this to "Name of the data source provider"

File: README.md [hunk 7]
```diff
@@ -87,3 +87,6 @@ An _example_ of a valid `data.json` file can be found in [`data.json`](src/data.
...
+- `source_url` - Source provider url
```
Comment (jessex) [inline README.md:91]: Let's change this to "URL of the data source provider"

File: README.md [hunk 8]
```diff
@@ -87,3 +87,6 @@ An _example_ of a valid `data.json` file can be found in [`data.json`](src/data.
...
+- `raw_source_categories` - Source categories
```
Comment (jessex) [inline README.md:92]: Let's change this to "Categories of data provided by the data source"

File: src/utils/generateSourceText.js [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+/**
+ * Takes source name and source categories and produces formatted string.
+ * @param sourceName {string}
+ * @param sourceCategories {string[]}
+ * @returns {string}
+ */
+const generateSourceText = (sourceName, sourceCategories) => {
+  if (sourceCategories.length) {
+    return `Includes data for the following categories from ${sourceName}'s public reports: ${sourceCategories.join(
+      ","
```
Comment (jessex) [inline src/utils/generateSourceText.js:26]: Whoops this got lost in the shuffle: put a space after the comma: `", "`

**Review by jessex — CHANGES_REQUESTED** [review]
Thanks, @Wmaileq! A few more comments for you. Also, I'll note that @hobuobi and @colincadams made a slight change in the tooltip on #31 that should be incorporated here before we merge.

Finally, I see a couple of remaining issues while testing, shown below.

**The text in the flow diagram can spill over in mobile layouts that are a bit thinner:**

Screen Shot 2021-01-29 at 15 28 47

Screen Shot 2021-01-29 at 15 28 59

**The math in the key insights cards feels off. The left-most one says a 31% drop on a large number of people but then says it's a reduction of only 10. The second card also seems off (if it's really a drop of 84 people then it should be more around ~40%). Are all of those numbers coming directly from the data file (the big current number, the change number, and the percentage change) or are any of those being calculated in the app? If they're being calculated here there may be a bug:**

Screen Shot 2021-01-29 at 15 29 29

**Comment by Wmaileq** [comment]
> The math in the key insights cards feels off. The left-most one says a 31% drop on a large number of people but then says it's a reduction of only 10. The second card also seems off (if it's really a drop of 84 people then it should be more around ~40%). Are all of those numbers coming directly from the data file (the big current number, the change number, and the percentage change) or are any of those being calculated in the app? If they're being calculated here there may be a bug:

This data comes directly from sources and is not calculated in the app.
I've looked through the data files and noticed that current version of data file consists of real and fake values. So I can assume that fake values were used here.

**Comment by jessex** [comment]
Thanks for letting me know, @Wmaileq! Sounds good on that point. However, you marked all of my comments from my last review as resolved even though they haven't been and my last review was more recent than your last commit. Did you perhaps make changes but forget to push the commit up to the repo? I just did another local test to be sure and the issues I called out in my most recent review are still apparent: my inline comments from the review, the mobile overflow issue, and the changes in the requirements on the issue that need to be taken into account.

**Comment by Wmaileq** [comment]
@jessex I am working on a dynamic font size change in dependence on window width.
I am trying to do a dynamic font size component to prevent text overflow.
Is it OK if boxes, where text overflows, will have a lesser font size than the others? Doesn't it feel off if font size will differ from card to card in dependence on content?
image

**Comment by hobuobi** [comment]
> Is it OK if boxes, where text overflows, will have a lesser font size than the others? Doesn't it feel off if font size will differ from card to card depending on the content?

That's a good point! If we reduced the font size for all of the text in the metric blocks, that might be a better solution (rather than trying to change it only if it overflows).

**Comment by jessex** [comment]
@Wmaileq Feel free to reduce text size across the board based on @hobuobi's comment above, but obviously it's going to be impossible to get a font size that works for all content and all screen resolutions, so let's just try to make a best effort that ensures that longer text (e.g. larger numbers or "Not available") don't overflow on mobile layouts.

**Review by jessex — COMMENTED** [review]
Great work, @Wmaileq! All that remains at this point is the text size issue that we're discussing in the flow diagram cards. But this appears to be good-to-go otherwise!

**Comment by Wmaileq** [comment]
Made card font size dynamic on mobile devices (320-439px)

**Review by jessex — APPROVED** [review]
The dynamic text sizing works great. Thanks, @Wmaileq!

**Pull request merged**